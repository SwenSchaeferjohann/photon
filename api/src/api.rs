use async_trait::async_trait;
use sea_orm::{DatabaseConnection, SqlxPostgresConnector};
use sqlx::{postgres::PgPoolOptions, Executor};

use crate::{
    error::PhotonApiError,
    method::{
        get_compressed_account::{
            get_compressed_account, GetCompressedAccountRequest, GetCompressedAccountResponse,
        },
        get_compressed_account_proof::{
            get_compressed_account_proof, GetCompressedAccountProofRequest,
            GetCompressedAccountProofResponse,
        },
        get_compressed_token_accounts_by_owner::{
            get_compressed_account_token_accounts_by_owner, GetCompressedTokenInfoByOwnerRequest,
            GetCompressedTokenInfoByOwnerResponse,
        },
    },
};

#[async_trait]
pub trait ApiContract: Send + Sync + 'static {
    async fn liveness(&self) -> Result<(), PhotonApiError>;

    async fn readiness(&self) -> Result<(), PhotonApiError>;

    async fn get_compressed_account(
        &self,
        payload: GetCompressedAccountRequest,
    ) -> Result<Option<GetCompressedAccountResponse>, PhotonApiError>;

    async fn get_compressed_account_proof(
        &self,
        payload: GetCompressedAccountProofRequest,
    ) -> Result<Option<GetCompressedAccountProofResponse>, PhotonApiError>;

    async fn get_compressed_account_token_accounts_by_owner(
        &self,
        payload: GetCompressedTokenInfoByOwnerRequest,
    ) -> Result<GetCompressedTokenInfoByOwnerResponse, PhotonApiError>;
}

pub struct PhotonApiConfig {
    pub db_url: String,
    pub max_conn: i32,
    pub timeout_seconds: i32,
}

pub struct PhotonApi {
    db_conn: DatabaseConnection,
}

impl PhotonApi {
    pub async fn new(config: PhotonApiConfig) -> Result<Self, anyhow::Error> {
        let PhotonApiConfig {
            db_url,
            max_conn,
            timeout_seconds,
            ..
        } = config;
        let db_conn = init_pool(&db_url, max_conn, timeout_seconds).await?;
        Ok(Self { db_conn })
    }
}

#[async_trait]
impl ApiContract for PhotonApi {
    async fn liveness(&self) -> Result<(), PhotonApiError> {
        Ok(())
    }

    async fn readiness(&self) -> Result<(), PhotonApiError> {
        todo!();
    }

    async fn get_compressed_account(
        &self,
        request: GetCompressedAccountRequest,
    ) -> Result<Option<GetCompressedAccountResponse>, PhotonApiError> {
        get_compressed_account(&self.db_conn, request).await
    }

    async fn get_compressed_account_proof(
        &self,
        request: GetCompressedAccountProofRequest,
    ) -> Result<Option<GetCompressedAccountProofResponse>, PhotonApiError> {
        get_compressed_account_proof(&self.db_conn, request).await
    }

    async fn get_compressed_account_token_accounts_by_owner(
        &self,
        request: GetCompressedTokenInfoByOwnerRequest,
    ) -> Result<GetCompressedTokenInfoByOwnerResponse, PhotonApiError> {
        get_compressed_account_token_accounts_by_owner(&self.db_conn, request).await
    }
}

async fn init_pool(
    db_url: &str,
    max_conn: i32,
    timeout_seconds: i32,
) -> Result<DatabaseConnection, sqlx::Error> {
    let pool = PgPoolOptions::new()
        .max_connections(max_conn as u32)
        .after_connect(move |conn, _meta| {
            Box::pin(async move {
                conn.execute(format!("SET statement_timeout = '{}s'", timeout_seconds).as_str())
                    .await?;
                Ok(())
            })
        })
        .connect(db_url)
        .await?;
    Ok(SqlxPostgresConnector::from_sqlx_postgres_pool(pool))
}
