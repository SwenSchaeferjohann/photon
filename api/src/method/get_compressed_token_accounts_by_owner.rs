use dao::generated::{token_owners};
use schemars::JsonSchema;
use sea_orm::{ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter};
use serde::{Deserialize, Serialize};

use crate::error::PhotonApiError;
use dao::typedefs::serializable_pubkey::SerializablePubkey;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetCompressedTokenInfoByOwnerRequest {
    pub owner: SerializablePubkey,
    pub mint: Option<SerializablePubkey>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct TokenUxto {
    pub owner: SerializablePubkey,
    pub mint: SerializablePubkey,
    pub amount: u64,
    pub delegate: Option<SerializablePubkey>,
    pub is_native: bool,
    pub close_authority: Option<SerializablePubkey>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetCompressedTokenInfoByOwnerResponse {
    pub total: i64,
    pub items: Vec<TokenUxto>,
}

fn _parse_model(token_owner: token_owners::Model) -> Result<TokenUxto, PhotonApiError> {
    Ok(TokenUxto {
        owner: token_owner.owner.try_into()?,
        mint: token_owner.mint.try_into()?,
        amount: token_owner.amount as u64,
        delegate: token_owner
            .delegate
            .map(SerializablePubkey::try_from)
            .transpose()?,
        is_native: token_owner.is_native.is_some(),
        close_authority: token_owner
            .close_authority
            .map(SerializablePubkey::try_from)
            .transpose()?,
    })
}

pub async fn get_compressed_account_token_accounts_by_owner(
    conn: &DatabaseConnection,
    request: GetCompressedTokenInfoByOwnerRequest,
) -> Result<GetCompressedTokenInfoByOwnerResponse, PhotonApiError> {
    let GetCompressedTokenInfoByOwnerRequest { owner, mint } = request;

    let mut filter = token_owners::Column::Owner.eq::<Vec<u8>>(owner.into());
    if let Some(m) = mint {
        filter = filter.and(token_owners::Column::Mint.eq::<Vec<u8>>(m.into()));
    }

    let result = token_owners::Entity::find()
        .filter(filter)
        .all(conn)
        .await?;

    let total = result.len() as i64;
    let items: Result<Vec<TokenUxto>, PhotonApiError> =
        result.into_iter().map(_parse_model).collect();
    let items = items?;

    Ok(GetCompressedTokenInfoByOwnerResponse { total, items })
}
