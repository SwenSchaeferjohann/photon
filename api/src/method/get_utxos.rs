use dao::generated::utxos;
use schemars::JsonSchema;
use sea_orm::{ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter};
use serde::{Deserialize, Serialize};

use crate::error::PhotonApiError;
use dao::typedefs::hash::Hash;
use dao::typedefs::serializable_pubkey::SerializablePubkey;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetUtxosRequest {
    pub owner: SerializablePubkey,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct Utxo {
    pub hash: Hash,
    pub account: Option<SerializablePubkey>,
    pub owner: SerializablePubkey,
    pub data: String,
    pub tree: Option<SerializablePubkey>,
    // TODO: Consider making lamports a u64.
    pub lamports: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetUtxosResponse {
    pub total: i64,
    pub items: Vec<Utxo>,
}

fn _parse_model(utxo: utxos::Model) -> Result<Utxo, PhotonApiError> {
    Ok(Utxo {
        hash: utxo.hash.try_into()?,
        account: utxo.account.map(SerializablePubkey::try_from).transpose()?,
        #[allow(deprecated)]
        data: base64::encode(utxo.data),
        owner: utxo.owner.try_into()?,
        tree: utxo.tree.map(|tree| tree.try_into()).transpose()?,
        lamports: utxo.lamports as u64,
    })
}

pub async fn get_utxos(
    conn: &DatabaseConnection,
    request: GetUtxosRequest,
) -> Result<GetUtxosResponse, PhotonApiError> {
    let owner = request.owner;

    let filter = utxos::Column::Owner
        .eq::<Vec<u8>>(owner.into())
        .and(utxos::Column::Spent.eq(false));
    let result = utxos::Entity::find().filter(filter).all(conn).await?;

    Ok(GetUtxosResponse {
        total: result.len() as i64,
        items: result
            .into_iter()
            .map(_parse_model)
            .collect::<Result<Vec<Utxo>, PhotonApiError>>()?,
    })
}
