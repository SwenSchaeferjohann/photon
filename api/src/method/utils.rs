use dao::generated::utxos;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use dao::typedefs::hash::Hash;
use dao::typedefs::serializable_pubkey::SerializablePubkey;

use crate::error::PhotonApiError;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct Utxo {
    pub hash: Hash,
    pub account: Option<SerializablePubkey>,
    pub owner: SerializablePubkey,
    pub data: String,
    pub tree: Option<SerializablePubkey>,
    pub lamports: u64,
    pub slot_created: u64,
}

pub fn parse_utxo_model(utxo: utxos::Model) -> Result<Utxo, PhotonApiError> {
    Ok(Utxo {
        hash: utxo.hash.try_into()?,
        account: utxo.account.map(SerializablePubkey::try_from).transpose()?,
        #[allow(deprecated)]
        data: base64::encode(utxo.data),
        owner: utxo.owner.try_into()?,
        tree: utxo.tree.map(|tree| tree.try_into()).transpose()?,
        lamports: utxo.lamports as u64,
        slot_created: utxo.slot_updated as u64,
    })
}
