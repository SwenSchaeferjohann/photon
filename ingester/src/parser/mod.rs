use borsh::BorshDeserialize;
use light_merkle_tree_event::Changelogs;
use log::info;
use psp_compressed_pda::event::PublicTransactionEvent;
use solana_sdk::pubkey::Pubkey;

use crate::{error::IngesterError, transaction_info::TransactionInfo};

use self::bundle::EventBundle;

pub mod bundle;
use solana_program::pubkey;

const ACCOUNT_COMPRESSION_PROGRAM_ID: Pubkey =
    pubkey!("5QPEJ5zDsVou9FQS3KCauKswM3VwBEBu4dpL9xTqkWwN");
const NOOP_PROGRAM_ID: Pubkey = pubkey!("noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV");

pub fn parse_transaction(tx: TransactionInfo) -> Result<Vec<EventBundle>, IngesterError> {
    info!("Parsing transaction: {}", tx);
    let mut event_bundles = Vec::new();
    for instruction_group in tx.instruction_groups {
        let mut ordered_intructions = Vec::new();
        ordered_intructions.push(instruction_group.outer_instruction);
        ordered_intructions.extend(instruction_group.inner_instructions);

        for (index, instruction) in ordered_intructions.iter().enumerate() {
            if index < ordered_intructions.len() - 2 {
                let next_instruction = &ordered_intructions[index + 1];
                let next_next_instruction = &ordered_intructions[index + 2];
                // We need to check if the account compression instruction contains a noop account to determine
                // if the instruction emits a noop event. If it doesn't then we want avoid indexing
                // the following noop instruction because it'll contain either irrelevant or malicious data.
                if ACCOUNT_COMPRESSION_PROGRAM_ID == instruction.program_id
                    && instruction.accounts.contains(&NOOP_PROGRAM_ID)
                    && next_instruction.program_id == NOOP_PROGRAM_ID
                    && next_next_instruction.program_id == NOOP_PROGRAM_ID
                {
                    let changelogs = Changelogs::deserialize(&mut next_instruction.data.as_slice())
                        .map_err(|e| {
                            IngesterError::ParserError(format!(
                                "Failed to deserialize Changelogs: {}",
                                e
                            ))
                        })?;
                    event_bundles.push(EventBundle::LegacyChangeLogEvent(changelogs));

                    let public_transaction_event = PublicTransactionEvent::deserialize(
                        &mut next_next_instruction.data.as_slice(),
                    )
                    .map_err(|e| {
                        IngesterError::ParserError(format!(
                            "Failed to deserialize PublicTransactionEvent: {}",
                            e
                        ))
                    })?;

                    event_bundles.push(EventBundle::LegacyPublicStateTransaction(
                        public_transaction_event.into(),
                    ));
                }
            }
        }
    }
    Ok(event_bundles)
}
