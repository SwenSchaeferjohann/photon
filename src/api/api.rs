use std::sync::Arc;

use async_trait::async_trait;
use sea_orm::{
    ConnectionTrait, DatabaseConnection, SqlxPostgresConnector, Statement,
};
use sqlx::{postgres::PgPoolOptions, Executor};

use super::{
    error::PhotonApiError,
    method::{
        get_compressed_account::{get_compressed_account, GetCompressedAccountRequest},
        get_compressed_account_proof::{
            get_compressed_account_proof, GetCompressedAccountProofRequest,
            GetCompressedAccountProofResponse,
        },
        get_compressed_token_accounts_by_owner::{
            get_compressed_account_token_accounts_by_owner, GetCompressedTokenInfoByOwnerRequest,
            GetCompressedTokenInfoByOwnerResponse,
        },
        get_utxo::{get_utxo, GetUtxoRequest},
        get_utxo_proof::{get_utxo_proof, GetUtxoProofRequest},
        get_utxos::{get_utxos, GetUtxosRequest, GetUtxosResponse},
        utils::{ProofResponse, Utxo},
    },
};

#[async_trait]
pub trait ApiContract: Send + Sync + 'static {
    async fn liveness(&self) -> Result<(), PhotonApiError>;

    async fn readiness(&self) -> Result<(), PhotonApiError>;

    async fn get_compressed_account(
        &self,
        payload: GetCompressedAccountRequest,
    ) -> Result<Utxo, PhotonApiError>;

    async fn get_compressed_account_proof(
        &self,
        payload: GetCompressedAccountProofRequest,
    ) -> Result<GetCompressedAccountProofResponse, PhotonApiError>;

    async fn get_compressed_account_token_accounts_by_owner(
        &self,
        payload: GetCompressedTokenInfoByOwnerRequest,
    ) -> Result<GetCompressedTokenInfoByOwnerResponse, PhotonApiError>;

    async fn get_utxos(&self, payload: GetUtxosRequest)
        -> Result<GetUtxosResponse, PhotonApiError>;

    async fn get_utxo(&self, request: GetUtxoRequest) -> Result<Utxo, PhotonApiError>;

    async fn get_utxo_proof(
        &self,
        payload: GetUtxoProofRequest,
    ) -> Result<ProofResponse, PhotonApiError>;
}

pub struct PhotonApiConfig {
    pub db_url: String,
    pub max_conn: i32,
    pub timeout_seconds: i32,
}

pub struct PhotonApi {
    db_conn: Arc<DatabaseConnection>,
}

impl PhotonApi {
    pub async fn new(config: PhotonApiConfig) -> Result<Self, anyhow::Error> {
        let PhotonApiConfig {
            db_url,
            max_conn,
            timeout_seconds,
            ..
        } = config;
        let db_conn = init_pool(&db_url, max_conn, timeout_seconds).await?;
        Ok(Self {
            db_conn: Arc::new(db_conn),
        })
    }
}

impl From<Arc<DatabaseConnection>> for PhotonApi {
    fn from(db_conn: Arc<DatabaseConnection>) -> Self {
        Self { db_conn }
    }
}

#[async_trait]
impl ApiContract for PhotonApi {
    async fn liveness(&self) -> Result<(), PhotonApiError> {
        Ok(())
    }

    async fn readiness(&self) -> Result<(), PhotonApiError> {
        self.db_conn
            .execute(Statement::from_string(
                self.db_conn.as_ref().get_database_backend(),
                "SELECT 1".to_string(),
            ))
            .await
            .map(|_| ())
            .map_err(Into::into)
    }

    async fn get_compressed_account(
        &self,
        request: GetCompressedAccountRequest,
    ) -> Result<Utxo, PhotonApiError> {
        get_compressed_account(&self.db_conn, request).await
    }

    async fn get_compressed_account_proof(
        &self,
        request: GetCompressedAccountProofRequest,
    ) -> Result<GetCompressedAccountProofResponse, PhotonApiError> {
        get_compressed_account_proof(&self.db_conn, request).await
    }

    async fn get_compressed_account_token_accounts_by_owner(
        &self,
        request: GetCompressedTokenInfoByOwnerRequest,
    ) -> Result<GetCompressedTokenInfoByOwnerResponse, PhotonApiError> {
        get_compressed_account_token_accounts_by_owner(&self.db_conn, request).await
    }

    async fn get_utxos(
        &self,
        request: GetUtxosRequest,
    ) -> Result<GetUtxosResponse, PhotonApiError> {
        get_utxos(&self.db_conn, request).await
    }

    async fn get_utxo(&self, request: GetUtxoRequest) -> Result<Utxo, PhotonApiError> {
        get_utxo(&self.db_conn, request).await
    }

    async fn get_utxo_proof(
        &self,
        request: GetUtxoProofRequest,
    ) -> Result<ProofResponse, PhotonApiError> {
        get_utxo_proof(&self.db_conn, request).await
    }
}

async fn init_pool(
    db_url: &str,
    max_conn: i32,
    timeout_seconds: i32,
) -> Result<DatabaseConnection, sqlx::Error> {
    let pool = PgPoolOptions::new()
        .max_connections(max_conn as u32)
        .after_connect(move |conn, _meta| {
            Box::pin(async move {
                conn.execute(format!("SET statement_timeout = '{}s'", timeout_seconds).as_str())
                    .await?;
                Ok(())
            })
        })
        .connect(db_url)
        .await?;
    Ok(SqlxPostgresConnector::from_sqlx_postgres_pool(pool))
}
