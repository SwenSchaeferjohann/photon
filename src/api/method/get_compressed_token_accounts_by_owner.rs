use schemars::JsonSchema;
use sea_orm::DatabaseConnection;
use serde::{Deserialize, Serialize};

use super::utils::{OwnerOrDelegate, TokenAccountListResponse};
use super::{super::error::PhotonApiError, utils::fetch_token_accounts};
use crate::dao::typedefs::serializable_pubkey::SerializablePubkey;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetCompressedTokenAccountsByOwnerRequest {
    pub owner: SerializablePubkey,
    pub mint: Option<SerializablePubkey>,
}

pub async fn get_compressed_token_accounts_by_owner(
    conn: &DatabaseConnection,
    request: GetCompressedTokenAccountsByOwnerRequest,
) -> Result<TokenAccountListResponse, PhotonApiError> {
    let GetCompressedTokenAccountsByOwnerRequest { owner, mint } = request;
    fetch_token_accounts(conn, OwnerOrDelegate::Owner(owner), mint).await
}
