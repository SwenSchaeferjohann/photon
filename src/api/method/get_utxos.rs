use crate::dao::generated::utxos;
use schemars::JsonSchema;
use sea_orm::{ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter};
use serde::{Deserialize, Serialize};

use super::super::error::PhotonApiError;
use crate::dao::typedefs::serializable_pubkey::SerializablePubkey;

use super::utils::{parse_utxo_model, Utxo};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetUtxosRequest {
    pub owner: SerializablePubkey,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetUtxosResponse {
    // TODO: Add cursor
    pub items: Vec<Utxo>,
}

pub async fn get_utxos(
    conn: &DatabaseConnection,
    request: GetUtxosRequest,
) -> Result<GetUtxosResponse, PhotonApiError> {
    let owner = request.owner;

    let filter = utxos::Column::Owner
        .eq::<Vec<u8>>(owner.into())
        .and(utxos::Column::Spent.eq(false));
    let result = utxos::Entity::find().filter(filter).all(conn).await?;

    Ok(GetUtxosResponse {
        items: result
            .into_iter()
            .map(parse_utxo_model)
            .collect::<Result<Vec<Utxo>, PhotonApiError>>()?,
    })
}
