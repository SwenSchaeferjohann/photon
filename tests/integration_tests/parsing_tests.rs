use function_name::named;
use photon::api::api::ApiContract;
use photon::api::method::get_compressed_token_accounts_by_owner::GetCompressedTokenAccountsByOwnerRequest;
use photon::api::method::get_utxos::GetUtxosRequest;
use photon::dao::typedefs::serializable_pubkey::SerializablePubkey;
use photon::ingester::parser::parse_transaction;

use crate::utils::*;
use insta::assert_json_snapshot;
use photon::api::method::utils::TokenUxto;
use serial_test::serial;
use solana_sdk::pubkey::Pubkey;

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_e2e_utxo_parsing(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name,
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let tx = cached_fetch_transaction(
        &setup,
        "2y27eTCZ53DiFubUqBtstcrFrDCvr1sqCJFYDnjFVuZrrCGXvQPfVVosBv7mYF3LeJRy73EiGzqPX2vWHDg4iRCk",
    )
    .await;

    let events = parse_transaction(&tx, 0).unwrap();
    assert_eq!(events.len(), 1);
    for event in events {
        persist_bundle_using_connection(setup.db_conn.as_ref(), event)
            .await
            .unwrap();
    }
    let utxos = setup
        .api
        .get_utxos(GetUtxosRequest {
            owner: "8uxi3FheruZNcPfq4WKGQD19xB44QMfUGuFLij9JWeJ"
                .try_into()
                .unwrap(),
        })
        .await
        .unwrap();

    assert_eq!(utxos.items.len(), 1);
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_e2e_token_mint(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name,
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let tx = cached_fetch_transaction(
        &setup,
        "kSEwLwJmJASMLycq52gKpNe7VtZdvuYaK9Vro2JZU4LHDt9GxgybWXtv7q9SRFvWVkCp7SEmWr51AdmKHD5N5wo",
    )
    .await;

    let events = parse_transaction(&tx, 0).unwrap();
    assert_eq!(events.len(), 1);
    for event in events {
        persist_bundle_using_connection(&setup.db_conn.as_ref(), event)
            .await
            .unwrap();
    }

    let owner = "GTP6qbHeRne8doYekYmzzYMTXAtHMtpzzguLg2LLNMeD";

    let token_accounts = setup
        .api
        .get_compressed_token_accounts_by_owner(GetCompressedTokenAccountsByOwnerRequest {
            owner: owner.try_into().unwrap(),
            mint: None,
        })
        .await
        .unwrap();

    assert_eq!(token_accounts.items.len(), 1);
    let token_utxo = token_accounts.items.get(0).unwrap();
    let expected_token_utxo = TokenUxto {
        owner: owner.try_into().unwrap(),
        mint: SerializablePubkey::try_from("GDvagojL2e9B7Eh7CHwHjQwcJAAtiMpbvCvtzDTCpogP").unwrap(),
        amount: 200,
        delegate: None,
        is_native: false,
        close_authority: None,
        frozen: false,
    };
    assert_eq!(token_utxo, &expected_token_utxo);
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_e2e_lamport_transfer(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let tx = cached_fetch_transaction(
        &setup,
        "2FWS4xBAHiZKfZ9gwcq6i1wqvoYX2acj49wAHoP6CtqshQaFjgnSxcMUPANPMU3q8qoESUacmYNHG9LCCh3X59fB",
    )
    .await;

    let events = parse_transaction(&tx, 0).unwrap();
    assert_eq!(events.len(), 1);
    for event in events {
        persist_bundle_using_connection(&setup.db_conn, event)
            .await
            .unwrap();
    }

    let owner1 = "A79DKmTDe8VzfRLti3wTi7mbJ9ENZtK7eBHtJ4QYCR2Y";
    let owner2 = "K75mYjtM3zrpBqBwEtQ1RsM39YULMMMoP1pm5ZQkF2h";

    for owner in [owner1, owner2] {
        let utxos = setup
            .api
            .get_utxos(GetUtxosRequest {
                owner: SerializablePubkey::from(Pubkey::try_from(owner).unwrap()),
            })
            .await
            .unwrap();
        assert_json_snapshot!(format!("{}-{}-utxos", name, owner), utxos);
    }
}
