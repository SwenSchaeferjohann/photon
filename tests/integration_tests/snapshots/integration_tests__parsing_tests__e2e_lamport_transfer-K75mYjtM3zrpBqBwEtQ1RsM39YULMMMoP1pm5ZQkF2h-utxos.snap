---
source: tests/integration_tests/parsing_tests.rs
expression: utxos
---
{
  "items": []
}
